patches/chromium/feat_add_frame-level_find_request_managers_in_web_contents_impl.patch

```diff
@@ -0,0 +1,419 @@
+ 
+@@ -286,18 +288,31 @@ void FindRequestManager::Find(int request_id,
+ }
+ 
+ void FindRequestManager::StopFinding(StopFindAction action) {
+-  for (WebContentsImpl* contents : contents_->GetWebContentsAndAllInner()) {
+-    for (FrameTreeNode* node : contents->GetFrameTree()->Nodes()) {
+-      RenderFrameHostImpl* rfh = node->current_frame_host();
+-      if (!CheckFrame(rfh) || !rfh->IsRenderFrameLive())
++  if (root_frame_) {
++    for (RenderFrameHost* frame: root_frame_->GetFramesInSubtree()) {
++      if (!CheckFrame(frame) || !frame->IsRenderFrameLive())
+         continue;
+       DCHECK(
+-          !static_cast<WebContentsImpl*>(WebContents::FromRenderFrameHost(rfh))
+-               ->IsPortal());
+-      rfh->GetFindInPage()->StopFinding(
++          !static_cast<WebContentsImpl*>(WebContents::FromRenderFrameHost(frame))
++              ->IsPortal());
++      static_cast<RenderFrameHostImpl*>(frame)->GetFindInPage()->StopFinding(
+           static_cast<blink::mojom::StopFindAction>(action));
+     }
+   }
++  else {
++    for (WebContentsImpl* contents : contents_->GetWebContentsAndAllInner()) {
++      for (FrameTreeNode* node : contents->GetFrameTree()->Nodes()) {
++        RenderFrameHostImpl* frame = node->current_frame_host();
++        if (!CheckFrame(frame) || !frame->IsRenderFrameLive())
++          continue;
++        DCHECK(
++            !static_cast<WebContentsImpl*>(WebContents::FromRenderFrameHost(frame))
++                ->IsPortal());
++        frame->GetFindInPage()->StopFinding(
++            static_cast<blink::mojom::StopFindAction>(action));
++      }
++    }
++  }
+ 
+   current_session_id_ = kInvalidId;
+ #if defined(OS_ANDROID)
+@@ -413,8 +428,9 @@ void FindRequestManager::RemoveFrame(RenderFrameHost* rfh) {
+   // removed, we can target any queued requests to the focused frame or main
+   // frame. However, if the main frame is removed we will not have a valid
+   // RenderFrameHost to target for the request queue.
+-  if (!rfh->GetParent())
++  if (!rfh->GetParent() || rfh == root_frame_) {
+     find_request_queue_ = base::queue<FindRequest>();
++  }
+ 
+   // Update the active match ordinal, since it may have changed.
+   if (active_frame_ == rfh) {
+@@ -596,8 +612,9 @@ void FindRequestManager::FindInternal(const FindRequest& request) {
+     // one, or the first frame with matches otherwise.
+     RenderFrameHost* target_rfh =
+         contents_->GetFocusedWebContents()->GetFocusedFrame();
+-    if (!target_rfh || !CheckFrame(target_rfh))
++    if (!target_rfh || !CheckFrame(target_rfh)) {
+       target_rfh = GetInitialFrame(request.options->forward);
++    }
+ 
+     SendFindRequest(request, target_rfh);
+     current_request_ = request;
+@@ -607,6 +624,21 @@ void FindRequestManager::FindInternal(const FindRequest& request) {
+ 
+   // This is an initial find operation.
+   Reset(request);
++
++  if (root_frame_) {
++    for (WebContentsImpl* contents : contents_->GetWebContentsAndAllInner()) {
```

Comment:
Shouldn't this just attach `FrameObserver` to the `webContents` associated with the `root_frame_`, why iterate over all of them ?